fuse-modules/encfs/encfs/encfs.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.3
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "ENCFS 1"
.TH ENCFS 1 "2006-03-08" "1.3.0" "Encrypted Filesystem"
.SH "NAME"
encfs \- mounts or creates an encrypted virtual filesystem
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBencfs\fR [\fB\-\-version\fR] [\fB\-s\fR] [\fB\-f\fR] [\fB\-v\fR|\fB\-\-verbose\fR] 
[\fB\-i \s-1MINUTES\s0\fR|\fB\-\-idle=MINUTES\fR] [\fB\-\-extpass=program\fR] 
[\fB\-S\fR|\fB\-\-stdinpass\fR] [\fB\-\-anykey\fR] [\fB\-\-forcedecode\fR] 
[\fB\-d\fR|\fB\-\-fuse\-debug\fR] [\fB\-\-public\fR] [\fB\-\-no\-default\-flags\fR]
\&\fIrootdir\fR \fImountPoint\fR 
[\fB\-\-\fR [\fIFuse Mount Options\fR]]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBencfs\fR creates a virtual encrypted filesystem which stores encrypted data in
the \fIrootdir\fR directory and makes the unencrypted data visible at the
\&\fImountPoint\fR directory.  The user must supply a password which is used to
(indirectly) encrypt both filenames and file contents.
.PP
If \fBencfs\fR is unable to find a supported filesystem at the specified
\&\fIrootdir\fR, then the user will be asked if they wish to create a new encrypted
filesystem at the specified location.  Options will be presented to the user
allowing some control over the algorithms to use.  As \fBEncFS\fR matures, there
may be an increasing number of choices.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-i\fR, \fB\-\-idle=MINUTES\fR" 4
.IX Item "-i, --idle=MINUTES"
Enable automatic unmount of the filesystem after a period of inactivity.  The
period is specified in minutes, so the shortest timeout period that can be
requested is one minute.  \fBEncFS\fR will not automatically unmount if there are
files open within the filesystem, even if they are open in read-only mode.
However simply having files open does not count as activity.
.IP "\fB\-f\fR" 4
.IX Item "-f"
The \fB\-f\fR (\fIforeground\fR) option causes \fBencfs\fR to run in the foreground.
Normally \fBencfs\fR spawns off as a daemon and runs in the background, returning
control to the spawning shell.  With the \fB\-f\fR option, it will run in the
foreground and any warning/debug log messages will be displayed on standard
error.  In the default (background) mode, all log messages are logged via
syslog.
.IP "\fB\-v\fR, \fB\-\-verbose\fR" 4
.IX Item "-v, --verbose"
Causes \fBencfs\fR to enable logging of various debug channels within \fBencfs\fR.
Normally these logging messages are disabled and have no effect.  It is
recommended that you run in foreground (\fB\-f\fR) mode when running with verbose
enabled.
.IP "\fB\-s\fR" 4
.IX Item "-s"
The \fB\-s\fR (\fIsingle threaded\fR) option causes \fBencfs\fR to run in single threaded
mode.  By default, \fBencfs\fR runs in multi-threaded mode.  This option is used
during \fBencfs\fR development in order to simplify debugging and allow it to run
under memory checking tools..
.IP "\fB\-d\fR, \fB\-\-fuse\-debug\fR" 4
.IX Item "-d, --fuse-debug"
Enables debugging within the \fB\s-1FUSE\s0\fR library.  This should only be used if you
suspect a problem within \fB\s-1FUSE\s0\fR itself (not \fBencfs\fR), as it generates a lot
of low-level data and is not likely to be very helpful in general problem
tracking.  Try \fIverbose\fR mode (\fB\-v\fR) first, which gives a higher level view
of what is happening within \fBencfs\fR.
.IP "\fB\-\-forcedecode\fR" 4
.IX Item "--forcedecode"
This option only has an effect on filesystems which use \s-1MAC\s0 block headers.  By
default, if a block is decoded and the stored \s-1MAC\s0 doesn't match what is
calculated, then an \s-1IO\s0 error is returned to the application and the block is
not returned.  However, by specifying \fB\-\-forcedecode\fR, only an error will be
logged and the data will still be returned to the application.  This may be
useful for attempting to read corrupted files.
.IP "\fB\-\-public\fR" 4
.IX Item "--public"
Attempt to make encfs behave as a typical multi-user filesystem.  By default,
all \s-1FUSE\s0 based filesystems are visible only to the user who mounted them.  No
other users (including root) can view the filesystem contents.  The \fB\-\-public\fR
option does two things.  It adds the \s-1FUSE\s0 flags \*(L"allow_other\*(R" and
\&\*(L"default_permission\*(R" when mounting the filesystem, which tells \s-1FUSE\s0 to allow
other users to access the filesystem, and to use the ownership permissions
provided by the filesystem.  Secondly, the \fB\-\-public\fR flag changes how encfs's
node creation functions work \- as they will try and set ownership of new nodes
based on the caller identification.
.Sp
\&\fBWarning\fR: In order for this to work, encfs must be run as root \*(-- otherwise
it will not have the ability to change ownership of files.  I recommend that
you instead investigate if the fuse allow_other option can be used to do what
you want before considering the use of \fB\-\-public\fR.
.IP "\fB\-\-\fR" 4
.IX Item "--"
The \fB\-\-\fR option tells \fBencfs\fR to send any remaining arguments directly to
\&\fB\s-1FUSE\s0\fR.  In turn, \fB\s-1FUSE\s0\fR passes the arguments to \fBfusermount\fR.  See
the \fBfusermount\fR help page for information on available commands.
.IP "\fB\-\-no\-default\-flags\fR" 4
.IX Item "--no-default-flags"
\&\fBEncfs\fR adds the \s-1FUSE\s0 flags \*(L"use_ino\*(R" and \*(L"default_permissions\*(R" by default, as
of version 1.2.2, because that improves compatibility with some programs..  If
for some reason you need to disable one or both of these flags, use the option
\&\fB\-\-no\-default\-flags\fR.
.Sp
The following command lines produce the same result:
.Sp
.Vb 2
\&    encfs raw crypt
\&    encfs --no-default-flags raw crypt -- -o use_ino,default_permissions
.Ve
.IP "\fB\-\-extpass=program\fR" 4
.IX Item "--extpass=program"
Specify an external program to use for getting the user password.  When the
external program is spawned, the environment variable \*(L"RootDir\*(R" will be set to
contain the path to the root directory.  The program should print the password
to standard output.
.Sp
\&\fBEncFS\fR takes everything returned from the program to be the password, except
for a trailing newline (\en) which will be removed.
.Sp
For example, specifying \fB\-\-extpass\fR=\fI/usr/lib/ssh/ssh\-askpass\fR will cause
\&\fBEncFS\fR to use ssh's password prompt program.  
.IP "\fB\-S\fR, \fB\-\-stdinpass\fR" 4
.IX Item "-S, --stdinpass"
Read password from standard input, without prompting.  This may be useful for
scripting encfs mounts.
.Sp
Note that you should make sure the filesystem and mount points exist first.
Otherwise encfs will prompt for the filesystem creation options, which may
interfere with your script.
.IP "\fB\-\-anykey\fR" 4
.IX Item "--anykey"
Turn off key validation checking.  This allows \fBEncFS\fR to be used with
secondary passwords.  This could be used to store a separate set of files in an
encrypted filesystem.  \fBEncFS\fR ignores files which do not decode properly, so
files created with separate passwords will only be visible when the filesystem
is mounted with their associated password.
.Sp
Note that if the primary password is changed (using \fBencfsctl\fR), the other
passwords will not be usable unless the primary password is set back to what it
was, as the other passwords rely on an invalid decoding of the volume key,
which will not remain the same if the primary password is changed.
.Sp
\&\fBWarning\fR: Use this option at your own risk.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
Create a new encrypted filesystem.  Store the raw (encrypted) data in
\&\*(L"~/.crypt\*(R" , and make the unencrypted data visible in \*(L"~/crypt\*(R".  Both
directories are in the home directory in this example.  This example shows the
full output of encfs as it asks the user if they wish to create the filesystem:
.PP
.Vb 9
\&    % encfs ~/.crypt ~/crypt
\&    Directory "/home/me/.crypt" does not exist, create (y,n)?y
\&    Directory "/home/me/crypt" does not exist, create (y,n)?y
\&    Creating new encrypted volume.
\&    Please choose from one of the following options:
\&     enter "x" for expert configuration mode,
\&     enter "p" for pre-configured paranoia mode,
\&     anything else, or an empty line will select standard mode.
\&    ?>
.Ve
.PP
.Vb 4
\&    Standard configuration selected.
\&    Using cipher Blowfish, key size 160, block size 512
\&    New Password: <password entered here>
\&    Verify: <password entered here>
.Ve
.PP
The filesystem is now mounted and visible in \fI~/crypt\fR.  If files are created
there, they can be seen in encrypted form in \fI~/.crypt\fR.  To unmount the
filesystem, use \fIfusermount\fR with the \fB\-u\fR (unmount) option:
.PP
.Vb 1
\&    % fusermount -u ~/crypt
.Ve
.PP
Another example.  To mount the same filesystem, but have fusermount name the
mount point '/dev/foo' (as shown in \fIdf\fR and other tools which read
/etc/mtab), and also request kernel-level caching of file data (which are both
special arguments to fusermount):
.PP
.Vb 1
\&    % encfs ~/.crypt ~/crypt -- -n /dev/foo -c
.Ve
.PP
Or, if you find strange behavior under some particular program when working in
an encrypted filesystem, it may be helpful to run in verbose mode while
reproducing the problem and send along the output with the problem report:
.PP
.Vb 1
\&    % encfs -v -f ~/.crypt ~/crypt 2> encfs-report.txt
.Ve
.PP
In order to avoid leaking sensitive information through the debugging channels,
all warnings and debug messages (as output in verbose mode) contain only
encrypted filenames.  You can use the \fIencfsctl\fR program's \fIdecode\fR function
to decode filenames if desired.
.SH "CAVEATS"
.IX Header "CAVEATS"
\&\fBEncFS\fR is not a true filesystem.  It does not deal with any of the actual
storage or maintenance of files.  It simply translates requests (encrypting or
decrypting as necessary) and passes the requests through to the underlying
host filesystem.  Therefor any limitations of the host filesystem will likely
be inherited by \fBEncFS\fR (or possibly be further limited).
.PP
One such limitation is filename length.  If your underlying filesystem limits
you to N characters in a filename, then \fBEncFS\fR will limit you to approximately
3*(N\-2)/4.  For example if the host filesystem limits to 256 characters, then
\&\fBEncFS\fR will be limited to 190 character filenames.  This is because encrypted
filenames are always longer then plaintext filenames.
.SH "FILESYSTEM OPTIONS"
.IX Header "FILESYSTEM OPTIONS"
When \fBEncFS\fR is given a root directory which does not contain an existing
\&\fBEncFS\fR filesystem, it will give the option to create one.  Note that options

fuse-modules/encfs/encfs/encfs.cpp
/*****************************************************************************
 * Author:   Valient Gough <vgough@pobox.com>
 *
 *****************************************************************************
 * Copyright (c) 2003-2004, Valient Gough
 * Modified 2006 Paolo Beverini
 * 
 * This program is free software; you can distribute it and/or modify it under 
 * the terms of the GNU General Public License (GPL), as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 */

#include "encfs.h"

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <dirent.h>
#include <errno.h>
#include <sys/statvfs.h>

#include <sys/types.h>
#ifdef linux
#include <sys/fsuid.h>
#endif

#ifdef HAVE_ATTR_XATTR_H
#include <attr/xattr.h>
#elif HAVE_SYS_XATTR_H
#include <sys/xattr.h>
#endif

#include <string>
#include <map>

#include "DirNode.h"
#include "MemoryPool.h"
#include "FileUtils.h"
#include <rlog/rlog.h>
#include <rlog/Error.h>

#ifndef MIN
#define MIN(a,b) (((a)<(b)) ? (a): (b))
#endif

using namespace std;
using namespace rlog;

struct EncfsFSContext
{
    void *reserved;//reserverd for main

    // for idle monitor
    bool running;
    pthread_t monitorThread;
    pthread_cond_t wakeupCond;
    pthread_mutex_t wakeupMutex;

    DirNode *FSRoot;
// if gPublicFilesystem is true, then try and make the filesystem behave like a
// normal multi-user filesystem (nodes are owned by the uid of the caller, etc)
    bool gPublicFilesystem;
    int oldStderr;
};


static RLogChannel *Info = DEF_CHANNEL("info", Log_Info);


/*
    The rLog messages below always print out encrypted filenames, not
    plaintext.  The reason is so that it isn't possible to leak information
    about the encrypted data through rlog interfaces.


    The purpose of this layer of code is to take the FUSE request and dispatch
    to the internal interfaces.  Any marshaling of arguments and return types
    can be done here.
*/

int encfs_getattr(const char *path, struct stat *stbuf)
{
    int res = -EIO;
    try
    {
        struct fuse_context *context=fuse_get_context();
        struct EncfsFSContext *ec=(struct EncfsFSContext *)context->private_data;

	Ptr<FileNode> fnode = ec->FSRoot->lookupNode( path, "getattr" );

	rLog(Info, "getattr on %s", fnode->cipherName());

	res = fnode->getAttr( stbuf );
    } catch( rlog::Error &err )
    {
	rError("error caught in getattr");
	err.log( _RLWarningChannel );
    }
    return res;
}

int encfs_getdir(const char *path, fuse_dirh_t h, fuse_dirfil_t filler)
{
    try
    {
	int res = 0;
        struct fuse_context *context=fuse_get_context();
        struct EncfsFSContext *ec=(struct EncfsFSContext *)context->private_data;

	DirTraverse dt = ec->FSRoot->openDir( path );

	rLog(Info, "getdir on %s", ec->FSRoot->cipherPath(path).c_str());

	if(dt.valid())
	{
	    int fileType = 0;
	    ino_t inode = 0;

	    std::string name = dt.nextPlaintextName( &fileType, &inode );
	    while( !name.empty() )
	    {
		res = filler( h, name.c_str(), fileType, inode );

		if(res != 0)
		    break;

		name = dt.nextPlaintextName( &fileType, &inode );
	    } 
	} else
	{
	    rInfo("getdir request invalid, path: '%s'", path);
	}

	return res;
    } catch( rlog::Error &err )
    {
	rError("Error caught in getdir");
	err.log( _RLWarningChannel );
	return -EIO;
    }
}

int encfs_mknod(const char *path, mode_t mode, dev_t rdev)
{
    int res = -EIO;
    try
    {
        struct fuse_context *context=fuse_get_context();
        struct EncfsFSContext *ec=(struct EncfsFSContext *)context->private_data;

	Ptr<FileNode> fnode = ec->FSRoot->lookupNode( path, "mknod" );

	rLog(Info, "mknod on %s, mode %i, dev %" PRIi64,
		fnode->cipherName(), mode, (int64_t)rdev);

	uid_t uid = 0;
	gid_t gid = 0;
	if(ec->gPublicFilesystem)
	{
	    uid = context->uid;
	    gid = context->gid;
	}
	res = fnode->mknod( mode, rdev, uid, gid );
    } catch( rlog::Error &err )
    {
	rError("error caught in mknod");
	err.log( _RLWarningChannel );
    }
    return res;
}

int encfs_mkdir(const char *path, mode_t mode)
{
    int res = -EIO;
    try
    {
        struct fuse_context *context=fuse_get_context();
        struct EncfsFSContext *ec=(struct EncfsFSContext *)context->private_data;

	uid_t uid = 0;
	gid_t gid = 0;
	if(ec->gPublicFilesystem)
	{
	    uid = context->uid;
	    gid = context->gid;
	}
	res = ec->FSRoot->mkdir( path, mode, uid, gid );
    } catch( rlog::Error &err )
    {
	rError("error caught in mkdir");
	err.log( _RLWarningChannel );
    }
    return res;
}

int encfs_unlink(const char *path)
{
    int res = -EIO;
    try
    {
        struct fuse_context *context=fuse_get_context();
        struct EncfsFSContext *ec=(struct EncfsFSContext *)context->private_data;

	// let DirNode handle it atomically so that it can handle race
	// conditions
	res = ec->FSRoot->unlink( path );
    } catch( rlog::Error &err )
    {
	rError("error caught in unlink");
	err.log( _RLWarningChannel );
    }
    return res;
}

int encfs_rmdir(const char *path)
{
    int res = -EIO;
    try
    {
        struct fuse_context *context=fuse_get_context();
        struct EncfsFSContext *ec=(struct EncfsFSContext *)context->private_data;

	string cyName = ec->FSRoot->cipherPath( path );

	rLog(Info, "rmdir %s", cyName.c_str());

	res = rmdir( cyName.c_str() );
	if(res == -1)
	{
	    int eno = errno;
	    rInfo("unlink error: %s", strerror(eno));
	    res = -eno;
	} else
	    res = 0;
    } catch( rlog::Error &err )
    {
	rError("error caught in rmdir");
	err.log( _RLWarningChannel );
    }
    return res;
}

int encfs_readlink(const char *path, char *buf, size_t size)
{
    try
    {
        struct fuse_context *context=fuse_get_context();
        struct EncfsFSContext *ec=(struct EncfsFSContext *)context->private_data;

	string cyName = ec->FSRoot->cipherPath( path );
	
	rLog(Info, "readlink %s", cyName.c_str());

	int res = ::readlink( cyName.c_str(), buf, size-1 );

	if(res == -1)
	    return -errno;

	buf[res] = '\0'; // ensure null termination
	string decodedName;
	try
	{
	    decodedName = ec->FSRoot->plainPath( buf );
	} catch(...) { }

	if(!decodedName.empty())
	{
	    strncpy(buf, decodedName.c_str(), size-1);
	    buf[size-1] = '\0';

	    return 0;
	} else
	{
	    rWarning("Error decoding link");
	    return -1;
	}
    } catch( rlog::Error &err )
    {
	rError("error caught in rmdir");
	err.log( _RLWarningChannel );
	return -EIO;
    }
}

int encfs_symlink(const char *from, const char *to)
{
    int res = -EIO;
    try
    {
        struct fuse_context *context=fuse_get_context();
        struct EncfsFSContext *ec=(struct EncfsFSContext *)context->private_data;

	// allow fully qualified names in symbolic links.
	string fromCName = ec->FSRoot->relativeCipherPath( from );
	string toCName = ec->FSRoot->cipherPath( to );
	
	rLog(Info, "symlink %s -> %s", fromCName.c_str(), toCName.c_str());

	// use setfsuid / setfsgid so that the new link will be owned by the
	// uid/gid provided by the fuse_context.
	int olduid = -1;
	int oldgid = -1;
	if(ec->gPublicFilesystem)
	{
	    olduid = setfsuid( context->uid );
	    oldgid = setfsgid( context->gid );
	}
	res = ::symlink( fromCName.c_str(), toCName.c_str() );
	if(olduid >= 0)
	    setfsuid( olduid );
	if(oldgid >= 0)
	    setfsgid( oldgid );

	if(res == -1)
	    res = -errno;
	else
	    res = 0;
    } catch( rlog::Error &err )
    {
	rError("error caught in symlink");
	err.log( _RLWarningChannel );
    }
    return res;
}

int encfs_link(const char *from, const char *to)
{
    int res = -EIO;
    try
    {
        struct fuse_context *context=fuse_get_context();
        struct EncfsFSContext *ec=(struct EncfsFSContext *)context->private_data;

	res = ec->FSRoot->link( from, to );
    } catch( rlog::Error &err )
    {
	rError("error caught in link");
	err.log( _RLWarningChannel );
    }
    return res;
}

int encfs_rename(const char *from, const char *to)
{
    int res = -EIO;
    try
    {
        struct fuse_context *context=fuse_get_context();
        struct EncfsFSContext *ec=(struct EncfsFSContext *)context->private_data;

	res = ec->FSRoot->rename( from, to );
    } catch( rlog::Error &err )
    {
	rError("error caught in rename");
	err.log( _RLWarningChannel );
    }
    return res;
}

int encfs_chmod(const char *path, mode_t mode)
{
    int res = -EIO;
    try
    {
        struct fuse_context *context=fuse_get_context();
        struct EncfsFSContext *ec=(struct EncfsFSContext *)context->private_data;

	string cyName = ec->FSRoot->cipherPath( path );

	rLog(Info, "chmod %s, mode %i", cyName.c_str(), mode);

	res = chmod( cyName.c_str(), mode );
	if(res == -1)
	{
	    int eno = errno;
	    rInfo("chmod error: %s", strerror(eno));
	    res = -eno;
	} else
	    res = 0;
    } catch( rlog::Error &err )
    {
	rError("error caught in chmod");
	err.log( _RLWarningChannel );
    }
    return res;
}

int encfs_chown(const char *path, uid_t uid, gid_t gid)
{
    int res = -EIO;
    try
    {
        struct fuse_context *context=fuse_get_context();
        struct EncfsFSContext *ec=(struct EncfsFSContext *)context->private_data;

	string cyName = ec->FSRoot->cipherPath( path );
	
	rLog(Info, "chown %s, uid %i, gid %i", cyName.c_str(), uid, gid);

	res = lchown( cyName.c_str(), uid, gid );
	if(res == -1)
	    res = -errno;
	else
	    res = 0;
    } catch( rlog::Error &err )
    {
	rError("error caught in chown");
	err.log( _RLWarningChannel );
    }
    return res;
}

int encfs_truncate(const char *path, off_t size)
{
    int res = -EIO;
    try
    {
        struct fuse_context *context=fuse_get_context();
        struct EncfsFSContext *ec=(struct EncfsFSContext *)context->private_data;

	Ptr<FileNode> fnode = ec->FSRoot->lookupNode( path, "truncate" );
       
	rLog(Info, "truncate %s", fnode->cipherName());
	
	res = fnode->truncate( size );
    } catch( rlog::Error &err )
    {
	rError("error caught in truncate");
	err.log( _RLWarningChannel );
    }
    return res;
}

int encfs_utime(const char *path, struct utimbuf *buf)
{
    int res = -EIO;
    try
    {
        struct fuse_context *context=fuse_get_context();
        struct EncfsFSContext *ec=(struct EncfsFSContext *)context->private_data;

	string cyName = ec->FSRoot->cipherPath( path );
	
	rLog(Info, "utime %s", cyName.c_str());

	res = utime( cyName.c_str(), buf);
	if(res == -1)
	{
	    int eno = errno;
	    rInfo("utime error: %s", strerror(eno));
	    res = -eno;
	} else
	    res = 0;
    } catch( rlog::Error &err )
    {
	rError("error caught in utime");
	err.log( _RLWarningChannel );
    }
    return res;
}

int encfs_open(const char *path, struct fuse_file_info *file)
{
    int res = -EIO;
    Ptr<FileNode> fnode;

    try
    {
        struct fuse_context *context=fuse_get_context();
        struct EncfsFSContext *ec=(struct EncfsFSContext *)context->private_data;

	fnode = ec->FSRoot->openNode( path, "open", file->flags, &res );

	if(!fnode.isNull())
	{
	    rLog(Info, "encfs_open for %s, flags %i", fnode->cipherName(), 
		    file->flags);

	    if( res >= 0 )
	    {
		file->fh = res; // store file handle
		res = 0; // success is 0
	    }
	}
    } catch( rlog::Error &err )
    {
	rError("error caught in open");
	err.log( _RLWarningChannel );
    }

    return res;
}

int encfs_flush(const char *path, struct fuse_file_info *file)
{
    (void)path;
    (void)file;
    return 0; // could also return -38 (function not implemented)..
}

/*
Note: This is advisory -- it might benefit us to keep file nodes around for a
bit after they are released just in case they are reopened soon.  But that
requires a cache layer.
 */
int encfs_release(const char *path, struct fuse_file_info *file)
{
    (void)file;
    int res = -EIO;
    try
    {
        struct fuse_context *context=fuse_get_context();
        struct EncfsFSContext *ec=(struct EncfsFSContext *)context->private_data;

	ec->FSRoot->release( path );

	res = 0;
    } catch( rlog::Error &err )
    {
	rError("error caught in release");
	err.log( _RLWarningChannel );
    }
    return res;
}

int encfs_read(const char *path, char *buf, size_t size, off_t offset,
	struct fuse_file_info *file)
{
    (void)file;
    int res = -EIO;
    try
    {
        struct fuse_context *context=fuse_get_context();
        struct EncfsFSContext *ec=(struct EncfsFSContext *)context->private_data;

	Ptr<FileNode> fnode = ec->FSRoot->lookupNode( path, "read" );

	rLog(Info, "read %s, offset %" PRIi64 ", size %i",
		fnode->cipherName(), offset, (int)size);

	ssize_t rdSz = fnode->read( offset, (unsigned char *)buf, size );

	res = rdSz;
    } catch( rlog::Error &err )
    {
	rError("error caught in read: %s", err.message());
	rError("read offset %" PRIi64 ", %i", offset, (int)size);
	err.log( _RLWarningChannel );
    }
    return res;
}

int encfs_fsync(const char *path, int dataSync,
	struct fuse_file_info *file)
{
    (void)path;
#ifndef linux
    (void)dataSync;
#endif

    int res = -EIO;
    try
    {
#ifdef linux
	if(dataSync == 0)
#endif
	    res = fsync( file->fh );
#ifdef linux
	else
	    res = fdatasync( file->fh );
#endif
	if(res == -1)
	    res = -errno;
    } catch( rlog::Error &err )
    {
	rError("error caught in fsync");
	err.log( _RLWarningChannel );
    }
    return res;
}

int encfs_write(const char *path, const char *buf, size_t size,
                     off_t offset, struct fuse_file_info *file)
{
    (void)file;
    int res = -EIO;
    try
    {
        struct fuse_context *context=fuse_get_context();
        struct EncfsFSContext *ec=(struct EncfsFSContext *)context->private_data;

	Ptr<FileNode> fnode = ec->FSRoot->lookupNode( path, "write" );

	rLog(Info, "write %s, offset %" PRIi64 ", size %i",
		fnode->cipherName(), offset, (int)size);

	if(fnode->write( offset, (unsigned char *)buf, size ))
	    res = size;
	else
	    rInfo("encfs_write: write failed");
    } catch( rlog::Error &err )
    {
	rError("error caught in write");
	err.log( _RLWarningChannel );
    }
    return res;
}

int encfs_statfs(const char *path, struct statvfs *st)
{
    int res = -EIO;
    try
    {
        struct fuse_context *context=fuse_get_context();
        struct EncfsFSContext *ec=(struct EncfsFSContext *)context->private_data;

	(void)path; // path should always be '/' for now..
	rAssert( st != NULL );
	string cyName = ec->FSRoot->rootDirectory();

	rLog(Info, "doing statfs of %s", cyName.c_str());
	res = statvfs( cyName.c_str(), st );
	if(!res) 
	{
	    // adjust maximum name length..
	    st->f_namemax     = 6 * (st->f_namemax - 2) / 8; // approx..
	}
	if(res == -1)
	    res = -errno;
    } catch( rlog::Error &err )
    {
	rError("error caught in statfs");
	err.log( _RLWarningChannel );
    }
    return res;
}

#ifdef HAVE_XATTR
int encfs_setxattr( const char *path, const char *name,
	const char *value, size_t size, int flags )
{
    int res = -EIO;
    try
    {
        struct fuse_context *context=fuse_get_context();
        struct EncfsFSContext *ec=(struct EncfsFSContext *)context->private_data;

	string cyName = ec->FSRoot->cipherPath( path );

	rLog(Info, "setxattr %s", cyName.c_str());

	res = ::setxattr( cyName.c_str(), name, value, size, flags );
	if(res == -1)
	    res = -errno;
    } catch( rlog::Error &err )
    {
	rError("error caught in setxattr");
	err.log( _RLWarningChannel );
    }
    return res;
}

int encfs_getxattr( const char *path, const char *name,
	char *value, size_t size )
{
    int res = -EIO;
    try
    {
        struct fuse_context *context=fuse_get_context();
        struct EncfsFSContext *ec=(struct EncfsFSContext *)context->private_data;

	string cyName = ec->FSRoot->cipherPath( path );

	rLog(Info, "getxattr %s", cyName.c_str());

	res = ::getxattr( cyName.c_str(), name, value, size );
	if(res == -1)
	    res = -errno;
    } catch( rlog::Error &err )
    {
	rError("error caught in getxattr");
	err.log( _RLWarningChannel );
    }
    return res;
}

int encfs_listxattr( const char *path, char *list, size_t size )
{
    int res = -EIO;
    try
    {
        struct fuse_context *context=fuse_get_context();
        struct EncfsFSContext *ec=(struct EncfsFSContext *)context->private_data;

	string cyName = ec->FSRoot->cipherPath( path );

	rLog(Info, "listxattr %s", cyName.c_str());

	res = ::listxattr( cyName.c_str(), list, size );
	if(res == -1)
	    res = -errno;
    } catch( rlog::Error &err )
    {
	rError("error caught in listxattr");
	err.log( _RLWarningChannel );
    }
    return res;
}

int encfs_removexattr( const char *path, const char *name )
{
    int res = -EIO;
    try
    {
        struct fuse_context *context=fuse_get_context();
        struct EncfsFSContext *ec=(struct EncfsFSContext *)context->private_data;

	string cyName = ec->FSRoot->cipherPath( path );

	rLog(Info, "removexattr %s", cyName.c_str());

	res = ::removexattr( cyName.c_str(), name );
	if(res == -1)
	    res = -errno;
    } catch( rlog::Error &err )
    {
	rError("error caught in removexattr");
	err.log( _RLWarningChannel );
    }
    return res;
}
#endif // HAVE_XATTR

